### DISCUSSION TITLE
can you add text to a PDF

### DISCUSSION THREAD

**Discussion by deadpengwen** [pypdfium2-team/pypdfium2#266] [General]
in pypdfium2 version 3 build there was an "insert_text()" method, i cant seem to find anything simular in the Version 4, is there a way to do this

Thanks

**Comment by mara004** [comment]
Unfortunately it had to be removed, see e.g. https://github.com/pypdfium2-team/pypdfium2/discussions/160#discussioncomment-4140123
The code was based on a user-contributed sample to support Hindi using Harfbuzz text shaping, but
* This was very complex and probably inefficient[^1] - pdfium has much simpler built-in functions for regular scripts
* The translation only accounted correctly for horizontal but not for vertical position, meaning it would only work for a subset of foreign scripts. IIRC there were more inaccuracies (e.g. more values not taken into account, mixup of APIs that should better be wrapped separately), and I was not sure how to fix this properly.

I'm hoping to add back a simpler `insert_text()` method based on pdfium's higher-level text API in the future.
If a contributor were able to fix and maintain the text shaping, it could also be added back as a separate function, say `insert_text_foreign()`.

Note, if you want exactly the [old implementation](https://github.com/pypdfium2-team/pypdfium2/blob/604c582cafd41e5d8904145076d738558bf1c445/src/pypdfium2/_helpers/page.py#L233C1-L285C71) for some reason, you can always copy it into your project. It's really just a helper around pdfium's raw APIs that are also exposed publicly by the library.

[^1]: due to several calls per char to two different C libraries
